engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/TypeRepresentation.java

```diff
@@ -0,0 +1,188 @@
+    @Override
+    public String toString() {
+      return "Any";
+    }
+  }
+
+  /**
+   * Represents the type that is associated with values (atoms) of a given type.
+   *
+   * <p>Instances that are assigned this type are built with one of the available constructors, but
+   * statically we do not necessarily know which one.
+   */
+  record AtomType(QualifiedName fqn, BindingsMap.Type typeDescription)
+      implements TypeRepresentation {
+    @Override
+    public String toString() {
+      return fqn.item();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj instanceof AtomType atomType) {
+        return fqn.equals(atomType.fqn);
+      } else {
+        return false;
+      }
+    }
+
+    @Override
+    public int hashCode() {
+      return fqn.hashCode() * 31;
+    }
+  }
+
+  record ArrowType(TypeRepresentation argType, TypeRepresentation resultType)
+      implements TypeRepresentation {
+    @Override
+    public String toString() {
+      return "(" + argType + " -> " + resultType + ")";
+    }
+  }
+
+  record SumType(List<TypeRepresentation> types) implements TypeRepresentation {
+    public SumType {
+      if (types.size() < 2) {
+        throw new IllegalArgumentException("Sum type must have at least 2 types");
+      }
+    }
+
+    @Override
+    public String toString() {
+      String repr =
+          types.stream()
+              .map(TypeRepresentation::toString)
+              .reduce((a, b) -> a + " | " + b)
+              .orElse("");
+      return "(" + repr + ")";
+    }
+
+    private Set<TypeRepresentation> asSet() {
+      return new HashSet<>(types);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj instanceof SumType sumType) {
+        return asSet().equals(sumType.asSet());
+      } else {
+        return false;
+      }
+    }
+
+    @Override
+    public int hashCode() {
+      return asSet().hashCode();
+    }
+  }
+
+  record IntersectionType(List<TypeRepresentation> types) implements TypeRepresentation {
+    public IntersectionType {
+      if (types.size() < 2) {
+        throw new IllegalArgumentException("Compound type must have at least 2 types");
+      }
+    }
+
+    @Override
+    public String toString() {
+      String repr =
+          types.stream()
+              .map(TypeRepresentation::toString)
+              .reduce((a, b) -> a + " & " + b)
+              .orElse("");
+      return "(" + repr + ")";
+    }
+  }
+
+  /**
+   * Represents a type object, i.e. an object that is an instance of a type's identity.
+   *
+   * <p>This object allows to call static methods on that type or create instances of this type
+   * using its constructors, which will be assigned the corresponding AtomType.
+   *
+   * @param name the qualified name of the type
+   * @param typeDescription the type description from the BindingsMap
+   */
+  record TypeObject(QualifiedName name, BindingsMap.Type typeDescription)
```

Comment:
Initially I was hesitant of keeping `BindingsMap.Type` inside of `TypeRepresentation` as it breaks the simplicity of the `TypeRepresentation` structure which would otherwise be 'self-contained'.

I have tried implementing an alternative approach where we only store the `QualifiedName` of the type and get to the type description based on the qualified name, resolving it from `PackageRepository`. However, I had trouble with that because in some raw compiler tests, the `PackageRepository` is simply not available. I could probably amend the testing infrastructure to make it available, but after a lot of back and forth, I decided that the main purpose of `TypeRepresentation` is to aid in type inference and not to be 'as simple as possible'. The logic of type inference is greatly simplified if I _keep_ this type description reference here, so that seemed like the most pragmatic solution.

Reply:
> The problem is not having `BindingsMap.Type` inside of `TypeObject`, but the fact that it is not encapsulated - it should be a `private` field (which implies the `TypeObject` should be a `class`, not `record`).

Hmmm, I can consider that but it feels counter-intuitive to me - I do need to use this `TypeObject` inside of `TypeInference` pass. If it is private, I either need to give it a public getter (but then it's the same as current state with `record`). Or do you mean to move the related logic to the inside of this class? I rather wanted the `TypeRepresentation` to be a thin data-container and the logic to be contained in `TypeInference` and its related helpers, but I'm open to suggestions. Can you please clarify what approach you meant to suggest here?

Reply:
> keeping BindingsMap.Type inside of TypeRepresentation as it breaks the simplicity of the TypeRepresentation structure

If you have to expose `BindingsMap.Type` from `TypeRepresentation`...

> ...move the related logic to the inside of this class? I rather wanted the TypeRepresentation to be a thin data-container and the logic to be contained in...

then it leaks out and `TypeRepresentation` isn't simple/thin.

Reply:
Thank you for the suggestion. 

In 91e0e1e03661b8647daeff9ef2fa4c4c67739f1c I have removed the direct `BindingsMap` reference from `TypeRepresentation` in favor of an abstract `AtomTypeInterface` encapsulating what the type inference needs to know about a type's "interface". It is still built on top of the `BindingsMap.Type` under the hood - so in terms of metadata serialization nothing changes really, but it conceptually decouples the `TypeRepresentation` from `BindingsMap`.

Reply:
Edit to above - updated commit hash to 789d722d3ec94dcd3419c191f4073169ee6f56f9 as I force pushed some forgotten changes.